feat: expose engine FFI API to a Dart consumer (prototype 3)

bassosimone opened the pull request:
The general idea here is to avoid returning -1, having a more
uniform API, and allowing for sync calls.

It's a bit more work on both sides but it's more regular, so maybe
this change is for the greater good?

This will be a third pull request because it varies from
https://github.com/ooni/probe-cli/pull/850 mostly in
the way in which we interface at the FFI level.

Reference issue: https://github.com/ooni/probe/issues/2197

bassosimone committed:
feat: expose engine FFI API to a Dart consumer

This diff is a variation of https://github.com/ooni/probe-cli/pull/849
where we use protobuf instead of a custom JSON message format
managed by us. Therefore, this diff allows us to evaluate which
is the cost of inventing a custom JSON data format.

At the FFI level, the most relevant change is that we are not
passing along UTF-8 strings anymore, rather we're now using
binary buffers. I spent some (but not too much) trying to see
whether we could actually avoid unnecessary copies. I could
do that for events but could not do that for messages. I think
it's fair enough since events are more frequent and we're not
exchanging super large messages anyway.

Also, the whole Dart codebase is a bit leaner because in the
meanwhile I leveled up a bit :^).

bassosimone committed:
Simplify the C API.

The general idea here is to avoid returning -1, having a more
uniform API, and allowing for sync calls.

It's a bit more work on both sides but it's more regular, so maybe
this change is for the greater good?

This will be a third pull request because it varies from
https://github.com/ooni/probe-cli/pull/850 mostly in
the way in which we interface at the FFI level.

bassosimone committed:
fix: automatically figure out which library to load

bassosimone committed:
fix: remove autogenerated files

bassosimone committed:
fix: ensure we can build DSOs for Android

bassosimone committed:
fix(dart): don't compare int against UINT32_MAX

This is quite likely to be surprising on 32 bit systems.

While there, improve the documentation.

ainghazal commented on dart/ooniengine/pubspec.lock:
```diff
@@ -0,0 +1,145 @@
+# Generated by pub
```
perhaps you want to ignore this one too?

bassosimone replied:
From https://dart.dev/tools/pub/glossary#lockfile

> A file named pubspec.lock that specifies the concrete versions and other identifying information for every immediate and transitive dependency a package relies on.
>
> Unlike the pubspec, which only lists immediate dependencies and allows version ranges, the lock file comprehensively pins down the entire dependency graph to specific versions of packages. A lockfile ensures that you can recreate the exact configuration of packages used by an application.
>
> The lockfile is generated automatically for you by pub when you run [pub get](https://dart.dev/tools/pub/cmd/pub-get), [pub upgrade](https://dart.dev/tools/pub/cmd/pub-upgrade), or [pub downgrade](https://dart.dev/tools/pub/cmd/pub-downgrade). If your package is an application package, you will typically check this into source control. For library packages, you usually won’t.

Because this is mostly a library, I'd say we want to commit it.

ainghazal commented on mk:
```diff
@@ -339,6 +339,80 @@ GOMOBILE = $(shell go env GOPATH)/bin/gomobile
 OONIMKALL_V := $(shell date -u +%Y.%m.%d-%H%M%S)
 OONIMKALL_R := $(shell git describe --tags || echo '0.0.0-dev')
 
+#help:
```
leaving here a couple alternatives to make that I've got in my radar (you probably know them already):

* https://github.com/fugue/zim/
* https://magefile.org/

ainghazal commented on pkg/ooniengine/geoip.go:
```diff
@@ -0,0 +1,60 @@
+// main implements taskRunner.main.
+func (r *geoIPRunner) main(ctx context.Context, emitter taskMaybeEmitter, args []byte) {
+	logger := newTaskLogger(emitter)
+	var config abi.GeoIPConfig
```
pretty obvious, but it seems like there's some boilerplate here in the session init that can be factored out :)

bassosimone committed:
feat: experimentally use cgo.NewHandle

It _should_ be equivalent to the code we had before in many respects
except less safe depending on how it's invoked by FFI.

However, it's simpler to write and less code for us to maintain.

So, let's use it for a while and see what happens.

bassosimone committed:
fix: ignore file containing dart packages

bassosimone committed:
feat: first attempt at creating iOS libraries

bassosimone committed:
fix: create xcframework for iOS

aanorbel committed:
Adds Flutter with sample usage (#880)

* Squashed 'dart/ooniprobe/' content from commit 88a93139

git-subtree-dir: dart/ooniprobe
git-subtree-split: 88a9313933335928e226a6745fefe85a316e75a4

* added system tray icon manager

* update implementation for windows platform

* added changes and linux specific configuration
